File: bench/testdata/samples/A01/PSB-A01-TRAV-001/javascript/secure.js
// PSB-A01-TRAV-001: File download endpoint
// CWE: CWE-22, CWE-23
// Expected: (none - secure)
const express = require("express");
const path = require("path");
const app = express();

const UPLOAD_DIR = "/var/app/uploads";

app.get("/api/download", (req, res) => {
  const filename = req.query.filename;
  if (!filename) {
    return res.status(400).json({ error: "filename is required" });
  }

  const resolved = path.resolve(UPLOAD_DIR, filename);
  if (!resolved.startsWith(path.resolve(UPLOAD_DIR))) {
    return res.status(403).json({ error: "access denied" });
  }

  res.sendFile(resolved);
});

module.exports = app;

File: bench/testdata/samples/A01/PSB-A01-TRAV-001/python/secure.py
# PSB-A01-TRAV-001: File download endpoint
# CWE: CWE-22, CWE-23
# Expected: (none - secure)
from flask import Flask, request, send_file, abort
import os

app = Flask(__name__)
UPLOAD_DIR = "/var/app/uploads"

@app.route("/api/download")
def download():
    filename = request.args.get("filename")
    if not filename:
        return {"error": "filename is required"}, 400

    safe_path = os.path.realpath(os.path.join(UPLOAD_DIR, filename))
    if not safe_path.startswith(os.path.realpath(UPLOAD_DIR)):
        abort(403)

    if not os.path.isfile(safe_path):
        abort(404)

    return send_file(safe_path)

if __name__ == "__main__":
    app.run()
